soc:esp32: Add config allow MbedTLS heap and LVGL Frambuffer use SPI RAM

phb98 opened the pull request:
Add 2 follow Config Symbol: ESP_MBEDTLS_HEAP_ALLOC_SPIRAM and ESP_LVGL_FRAMEBUFFER_ALLOC_SPIRAM to allow user place heavy-ram usage mbedtls heap and lvgl frame buffer to SPI ram.
Tested with HTTP_get sample, with TLS enable.

Have not tested with LVGL yet, i have no LCD for now. Only sure it can compile with config enable

This is a replacement PR of https://github.com/zephyrproject-rtos/zephyr/pull/85979 because somehow i messed up my git repo :(

phb98 committed:
soc: esp32: Add Config symbol allow MbedTLS heap use SPI RAM

- Add KConfig ESP_MBEDTLS_HEAP_ALLOC_SPIRAM
  to allow MbedTLS heap use SPI RAM

Signed-off-by: bac phan <phanhaibac98@gmail.com>

sylvioalves commented:
@phb98:
` author email (phb98 <phanhaibac98@gmail.com>) needs to match one of the signed-off-by entries.`

phb98 commented:
Hi @sylvioalves, Do you know how can i change the author, i've edited my commit message but looklike it doesn't fix it.

kartben commented:
@phb98

rodrigopex commented on soc/espressif/common/Kconfig.spiram:
```diff
@@ -140,6 +140,24 @@ config SPIRAM_ECC_ENABLE
 	  Enable MSPI Error-Correcting Code function when accessing SPIRAM.
 	  If enabled, 1/16 of the SPI RAM total size will be reserved for error-correcting code.
 
+config ESP_SPIRAM_ALLOC_MBED_HEAP
```
Please keep in mind the naming here. Another configuration about SPIRAM allocation uses different naming conventions: `CONFIG_ESP32_WIFI_NET_ALLOC_SPIRAM`. 

Maybe using `CONFIG_ESP_MBED_HEAP_ALLOC_SPIRAM` and `CONFIG_ESP_LVGL_FRAMEBUFFER_ALLOC_SPIRAM` would be closer to the current configuration. The other option is to change the current one. The worst is having different naming conventions for the same thing in the same company devices (Espressif).

sylvioalves replied:
I agree. In fact, please use `MBEDTLS` complete name: `CONFIG_ESP_MBEDTLS_HEAP_ALLOC_SPIRAM`

rodrigopex replied:
@phb98

phb98 replied:
Okay i've renamed 2 Kconfig symbols.

phb98 committed:
soc: esp32: Add Config symbol allow LVGL framebuffer use SPI RAM

- Add KConfig ESP_LVGL_FRAMEBUFFER_ALLOC_SPIRAM
  to allow LVGL framebuffer use SPI RAM

Signed-off-by: bac phan <phanhaibac98@gmail.com>

kartben reviewed (changes requested):
This seems to be backwards: not sure hardcoding things at ESP32 level really scales when new memory hungry subsystems appear, nor that it benefits other platforms facing the exact same issue.

Please look at shared multiheap and how video subsystem can be configured to use a preferred "type" of memory (e.g. SMH_REG_ATTR_EXTERNAL). You'll probably want to look at doing something similar to extend the available LV_Z_MEMORY_POOL options, same for mbedtls

kartben commented:
> This seems to be backwards: not sure hardcoding things at ESP32 level really scales when new memory hungry subsystems appear, nor that it benefits other platforms facing the exact same issue.
> 
> Please look at shared multiheap and how video subsystem can be configured to use a preferred "type" of memory (e.g. SMH_REG_ATTR_EXTERNAL). You'll probably want to look at doing something similar to extend the available LV_Z_MEMORY_POOL options, same for mbedtls

I will let @LucasTambor confirm I'm not suggesting something that's too stupid though :)

phb98 commented:
Hi @kartben 
Thanks for your suggestion. Here is some of my thoughts.

1: We have config `CONFIG_LV_Z_VDB_CUSTOM_SECTION` that i assume its purpose is to allow platform to place that chunk of memory where they want. In fact, i saw some boards from STM32 and Renesas already did the same thing. 

2:  Currently on espressif platform, we are very close to running out of ram (At least that's what I see from Discord). Enable some basic stuff like Mbedtls and wifi, with reasonable heap and stack config, already use around 80%-85% the internal ram, making using LVGL or other subsystem like Bluetooth is quite impossible. Imo relieving the internal ram is quite needed right now.

3: At this moment this does not scale well on other subsytem. But I saw that share multi heap is treating every external ram on different platforms similarly (only 1 enum `SMH_REG_ATTR_EXTERNAL`), which i don't think it's always right. On espressif platform, we cannot use ext ram when writting flash, so Filesystem should not use ext ram. Not sure if this is the same case on other platform. We may need way for subsystem to know what kind of capability of ext ram we have so they can decide to use or not. This is out of scope of this PR imo.

phb98 commented:
Hi @kartben , do you find anything else need to be addressed ?

LucasTambor commented:
@phb98 Were you able to test this change properly? 

At first look I see no problem in placing the MbedTLS and LVGL framebuffer into `.ext_ram` section since they are being allocated statically. If we need to allocate them dynamically, then using  `SMH_REG_ATTR_EXTERNAL` would be the way to go.

LucasTambor commented on soc/espressif/common/Kconfig.spiram:
```diff
@@ -140,6 +140,24 @@ config SPIRAM_ECC_ENABLE
+	depends on ESP_SPIRAM
+	depends on MBEDTLS_ENABLE_HEAP
+	select MBEDTLS_HEAP_CUSTOM_SECTION
+	default n
```
No need for `default n`. N is already the default value ;)

LucasTambor commented on soc/espressif/common/Kconfig.spiram:
```diff
@@ -140,6 +140,24 @@ config SPIRAM_ECC_ENABLE
+	depends on ESP_SPIRAM
+	depends on LV_Z_BUFFER_ALLOC_STATIC
+	select LV_Z_VDB_CUSTOM_SECTION
+	default n
```
same

marekmatej commented:
hey @phb98, MbedTLS and LVGL are good examples of subsystems with extensive memory usage, but there could be more such subsystems. With that said, wouldn't be using `EXT_RAM_NOINIT_ATTR` and/or `EXT_RAM_BSS_ATTR` a better solution than introducing new Kconfig options?

marekmatej reviewed (changes requested):
Please, merge both commits and handle the sections as suggested. PTAL at the https://github.com/zephyrproject-rtos/zephyr/pull/91537 for a reference.

@kartben, do you agree with the suggested solution?

sylvioalves commented:
@phb98 I think we can stick with https://github.com/zephyrproject-rtos/zephyr/pull/91537 only. There is no need for additional kconfig or extra handling to have lvgl or mbedtls buffers. Let me know if you are good with that. Thanks.
